### PR TITLE
Adds paletteID and switch() to more widgets

### PR DESCRIPTION
**Pull request by mlzummo**
@GordonSmith @jbrundage - Please Review 

Note: the D3 Line/Pie/Bubble paletteID in dermatology does not update (only the cloned one does any ideas?)

Signed-off-by: Mathew Zummo <mzummo@gmail.com>

### CODE REVIEW

**Comment by GordonSmith** [comment]
The fill is probably being done in the "enter" cycle and not the "update" cycle...

**Comment by mlzummo** [comment]
@GordonSmith 

Fixed widgets ... SunburstPartition though in general (on master) ... does not render its clone in dermatology, i am getting a requirejs error too ....

**Comment by GordonSmith** [comment]
That sunburst one is fixed I believe (awaiting review)

File: widgets/src/c3/Common1D.js [hunk 1]
```diff
@@ -26,5 +26,7 @@
     Common1D.prototype = Object.create(Common.prototype);
     Common1D.prototype.implements(I1DChart.prototype);
 
+    Common1D.prototype.publish("paletteID", "default", "set", "Palette ID", Common1D.prototype._palette.switch());
```
Comment (GordonSmith) [inline widgets/src/c3/Common1D.js]: This should not be here as you can't know the correct palette type

Comment (GordonSmith) [inline widgets/src/c3/Common1D.js]: I think this is breaking the Gauge?

File: widgets/src/c3/Gauge.js [hunk 2]
```diff
@@ -32,8 +32,7 @@
     Gauge.prototype.publish("arc_width", 50, "number", "Gauge width of arc");
     Gauge.prototype.publish("show_labels", true, "boolean", "Show Labels");
     Gauge.prototype.publish("show_value_label", true, "boolean", "Show Value Label");
-    Gauge.prototype.publish("paletteID", "RdYlGn", "set", "Palette ID", Gauge.prototype._palette.switch());
-
```
Comment (GordonSmith) [inline widgets/src/c3/Gauge.js]: I think this change breaks the Gauge?

File: widgets/src/chart/Bubble.js [hunk 3]
```diff
@@ -77,6 +81,7 @@
                 var pos = { x: d.x - context._size.width / 2, y: d.y - context._size.height / 2 }
                 element.select("circle").transition()
                     .attr("transform", function (d) { return "translate(" + pos.x + "," + pos.y + ")"; })
+                    .style("fill", function (d) { return context._palette(d[0]); })
```
Comment (GordonSmith) [inline widgets/src/chart/Bubble.js]: You should remove it from the enter otherwise it gets done twice.

File: widgets/src/chart/Line.js [hunk 4]
```diff
@@ -37,7 +41,9 @@
         ;
 
         line.enter().append("path")
-            .attr("class", "dataLine")
+            .attr("class", "dataLine");
```
Comment (GordonSmith) [inline widgets/src/chart/Line.js]: Put ";" on next line for code consistency.

File: widgets/src/chart/Pie.js [hunk 5]
```diff
@@ -27,10 +27,13 @@
             .outerRadius(this._radius)
             .innerRadius(this._innerRadius)
         ;
+
```
Comment (GordonSmith) [inline widgets/src/chart/Pie.js]: Blank line added by random

File: widgets/src/chart/Pie.js [hunk 6]
```diff
@@ -119,6 +123,7 @@
                 var element = d3.select(this);
                 element.select("path").transition()
                     .attr("d", context.d3Arc)
+                    .style("fill", function (d) { return context._palette(d.data[0]); })
```
Comment (GordonSmith) [inline widgets/src/chart/Pie.js]: You should remove it from the enter otherwise it gets done twice.

File: widgets/src/tree/SunburstPartition.js [hunk 7]
```diff
@@ -13,7 +13,9 @@
     };
     SunburstPartition.prototype = Object.create(SVGWidget.prototype);
     SunburstPartition.prototype.implements(ITree.prototype);
-
+	
+    //SunburstPartition.prototype.publish("paletteID", "default", "set", "Palette ID", SunburstPartition.prototype._palette.switch());
+
```
Comment (GordonSmith) [inline widgets/src/tree/SunburstPartition.js]: Commented out code - no need to commit.

File: widgets/src/tree/SunburstPartition.js [hunk 8]
```diff
@@ -43,7 +45,8 @@
 
     SunburstPartition.prototype.update = function (domNode, element) {
         var context = this;
-
+		
+        //this._palette = this._palette.switch(this._paletteID);
```
Comment (GordonSmith) [inline widgets/src/tree/SunburstPartition.js]: Commented out code.

File: widgets/src/tree/Dendrogram.js [hunk 9]
```diff
@@ -52,15 +55,17 @@
 
         //  Nodes  ---
         var nodes = this.g.selectAll(".node").data(nodes);
-        var node = nodes.enter().append("g")
+        var node_enter = nodes.enter().append("g")
             .attr("class", "node")
         ;
+        var circle = node_enter.append("circle");
```
Comment (GordonSmith) [inline widgets/src/tree/Dendrogram.js]: "var circle" not used

File: widgets/src/tree/Dendrogram.js [hunk 10]
```diff
@@ -52,15 +55,17 @@
...
+        var text = node_enter.append("text");
```
Comment (GordonSmith) [inline widgets/src/tree/Dendrogram.js]: "var text" not used.

File: widgets/src/c3/Common1D.js (hunk 1 above)
Comment (mlzummo) [inline widgets/src/c3/Common1D.js]: last commit fixes

**Comment by mlzummo** [comment]
@GordonSmith - Where can i see/test changes for sunburstpartition derm. fix?

**Comment by GordonSmith** [comment]
I think its here (waiting for review):  https://github.com/hpcc-systems/Visualization/pull/217

File: widgets/src/c3/Common1D.js (hunk 1 above)
Comment (GordonSmith) [inline widgets/src/c3/Common1D.js]: But guage is working as is?  I am not sure why your adding more code?

File: widgets/src/chart/Bubble.js [hunk 11]
```diff
@@ -35,10 +37,12 @@
     Bubble.prototype.update = function (domNode, element) {
         var context = this;
 
+        this._palette = this._palette.switch(this._paletteID);
         var node = element.selectAll(".node")
             .data(this._data.length ? this.d3Pack.nodes({ children: this.cloneData() }).filter(function (d) { return !d.children; }) : [], function (d) { return d[0]; })            
         ;
 
+
```
Comment (GordonSmith) [inline widgets/src/chart/Bubble.js]: Blank line

**Comment by mlzummo** [comment]
the problem is here:

dermatology.html:191 Uncaught TypeError: Converting circular structure to JSON
text.js:308 XHR finished loading: GET "http://rawgit.com/GordonSmith/Visualization/TREE_CLICK/widgets/src/tree/SunburstPartition.js".
require.js:1903 GET http://rawgit.com/GordonSmith/Visualization/TREE_CLICK/widgets/src/tree/Sunburst.js 
require.js:166 Uncaught Error: Script error for: src/tree/Sunburst
http://requirejs.org/docs/errors.html#scripterror

File: widgets/src/c3/Gauge.js [hunk 12]
```diff
@@ -19,12 +18,11 @@
             clickEvent[d.id] = d.value;
             context.click(clickEvent, d.id);
         };
-        this._config.data.color = function (color, d) {
+    	this._config.data.color = function (color, d) {
```
Comment (GordonSmith) [inline widgets/src/c3/Gauge.js]: indentation

**Comment by GordonSmith** [comment]
Just the one indentation issue, by the looks of it.

**Comment by GordonSmith** [comment]
Looks fine.

**Pull request merged**